### DISCUSSION TITLE
Difference in Grain Count and Grain IDs When Loading DREAM3D Files with DAMASK and h5py

### DISCUSSION THREAD

**Discussion by hugy888** [damask-multiphysics/DAMASK#8] [Q&A] [answered]
Hi,
I've encountered an issue when loading DREAM3D files of polycrystals. I used damask.Grid.load_DREAM3D() to load the files, and I noticed that the number of grains (# materials: 1143) identified by DAMASK is different from the count I get when using h5py (# materials: 1164). Could someone explain why DAMASK and h5py might be reporting different grain counts?

Furthermore, there seems to be a discrepancy in the grain ID arrays obtained from these two methods. The grain_id_damask array starts at 0, while the grain_id_h5read from h5py starts at 1. During the loading or simulation process, does DAMASK assign the first material from material.yaml as grain_id=0? Also, how does DAMASK assign each material to a grain ID—is it based on the order in material.yaml?

I've attached a snippet of my code and the outputs for reference:
![Untitled](https://github.com/eisenforschung/DAMASK/assets/60995897/b13bf6ea-fef9-4ac2-802f-e6aa9fe07fee)

I appreciate any insights or guidance on this matter.
Thanks

**Comment by sharanroongta** [comment]
Hi,

First I try to answer the second part of your question: Grain ID does start from 0, and it is mapped to the material list in material.yaml. So you have understood this correctly.

As far as the first query is concerned: 
Maybe @vitesh13 can provide some insight, who has more experience with DREAM3D data.

Hope this helps.
Thanks

**Comment by vitesh13** [comment, marked as answer]
Hi,

I just tried out the load_DREAM3D function in the latest version and it works fine. I get the correct grain IDs.

To help us understand the problem and test the code, it would be great if you can mention the DAMASK version that you are using and also attach the Dream3D file which is raising this error.

**Reply by hugy888** [reply]
Hi,

Thank you for your response. I am using DAMASK version 3.0.0a7.post0, and I have encountered this error when loading my Dream3D file. As requested, I've attached the Dream3D file for your reference and testing.

[Dream3d_files.zip](https://github.com/eisenforschung/DAMASK/files/14613478/Dream3d_files.zip)

Additionally, I attempted to use the newer DAMASK version 3.0.0b0 to see if the issue would resolve, but I encountered the same error. The details are as follows:

![image](https://github.com/eisenforschung/DAMASK/assets/60995897/5ddd464a-2f1f-4c21-8aba-fead82124f67)

I appreciate your assistance in resolving this matter.

Best regards.

**Reply by vitesh13** [reply]
Hi,
Thanks for the files. I could recreate the error that you had. 
One suggestion that solves it is that you specify the name of the dataset that contains FeatureIds:
`grid_damask = damask.GeomGrid.load_DREAM3D('equi_comp_03.dream3d',feature_IDs='FeatureIds')`

Then the material values are correctly displayed. Can you try this out and let us know if that works for you? I think that this should work with the latest version and also with version 3.0.0.alpha7. 

As to why the featureIds are not considered in the earlier attempts in comparison to example I tried on my own, I would have to investigate it a bit. I will post about it, if I get any further insights there. 

However, the solution above follows the example given in the help of the function. So this is still a standard usage.

**Reply by hugy888** [reply]
Thanks for your response and suggestion. I've tried specifying the name of the dataset as you mentioned, and indeed, the grain counts match perfectly in both versions of DAMASK. 

However, there are still differences in the two numpy arrays regarding the grain IDs, as illustrated below:

![image](https://github.com/eisenforschung/DAMASK/assets/60995897/0955e243-86ed-4bd9-b538-9ff7425c0f5a)

Interestingly, the bincount results of these arrays are identical, suggesting that the number of voxels for each grain ID is the same across both methods. This leads me to think that perhaps a rotation or some other transformation is applied during the Dream3D file loading process, since the grain size distribution for each grain ID corresponds exactly.

![image](https://github.com/eisenforschung/DAMASK/assets/60995897/3da1605e-6195-442f-9c4e-d5ce6b446fee)

Could there be a transformation step that I might be overlooking? Any further insights on this would be greatly appreciated.

Best regards.

**Reply by eisenlohr** [reply]
@hugy888

> [!IMPORTANT]
> Please _always_ paste __editable code snippets__ instead of screen shots.
This makes reusing and debugging your code much easier for the next person.

Example based on the screenshot above:
```python
comparison_bincount = np.bincount(grain_id_damask.reshape(-1)) == np.bincount(grain_id_h5read.reshape(-1))
print(comparison_bincount.all())
```

**Reply by eisenlohr** [reply]
I suspect that you are observing the difference in indexing order between DAMASK and Dream3D. Please try
```python
(grain_id_damask.flatten('F') == grain_id_h5read.flatten('C')).all()
```
DAMASK sorts with "x" being the fastest index, i.e. Fortran-style or column-major order, whereas Dream3D likely uses "z" as fastest index, i.e. C-style or row-major order (which happens to be the default for `flatten` and, equivalently, `reshape(-1)`).

> [!CAUTION]
> Hence, when feeding a flattened data array into a visualization pipeline, say `pyvista` or such, you need to be certain that the order you are providing is the order that is expected. Otherwise, you are inadvertently switching the "x" and "z" coordinate axes, resulting in a geometry inversion!

---

### Example using PyVista

```python
import numpy as np
import pyvista as pv
pv.set_jupyter_backend('client')

dim = np.array([5,2,3])                             # x, y, z
ids = np.arange(dim.prod()).reshape(dim,order='F')  # with order='F', ids[:,0,0] == array([0, 1, 2, 3, 4])
msh = pv.ImageData(dimensions=dim+1)                # 'dimensions' is number of points = number of cells+1
msh['id'] = ids.flatten('F')                        # pyvista expects "x" as fastest index and "z" as slowest
msh.plot()
```

[<img src="https://github.com/eisenforschung/DAMASK/assets/6528644/ce6d9900-230b-495a-9548-f9e21eb32c2d">](https://www.dropbox.com/scl/fi/xt74ujgn6qerej8ihwz9b/5x2x3.html?rlkey=g1mlxh6xi1jf24iuho08wb691&dl=0)

**Reply by eisenlohr** [reply]
Another way to look at this is to realize that:
```Python
>>> (grid_damask.material.flatten('F')[:128] == grid_damask.material[:,0,0]).all()
True
>>> (grid_damask.material.flatten('C')[:128] == grid_damask.material[0,0,:]).all()
True
```

**Reply by MarDiehl** [reply]
To explain the original observation: If “feature_Ids" is not given in the `load_DREAM3D` call, the grains identified by DREAM.3D through segmentation are not considered but cell-wise data is used, i.e. each cell is essentially its own “grain”. Nevertheless, to keep the material configuration files small, `load_DREAM3D` lumps any cells that have exactly the same orientation and phase into separate groups that then form a grain each. For synthetic data without any in-grain scatter, this strategy results in more or less the same number of materials (grain IDs) as when using the data from a proper DREAM3D segmentation.

@hugy888: Please mark this answer as solved if you have no further questions
